source/extensions/quic_listeners/quiche/platform/quic_server_stats_impl.h

```diff
@@ -6,18 +6,184 @@
+//          Envoy::Stats::Metric& d_() { return d; }
+//          EnumMap e;
+//          EnumMap& e_() { return e; }
+//        } c;
+//      } b;
+//    } a;
+//
+//    // QUIC_SERVER_HISTOGRAM_COUNTER(aaa, ...)
+//    Envoy::Stats::Histogram& aaa;
+//    Envoy::Stats::Histogram& aaa_() { return aaa; }
+//
+//    // QUIC_SERVER_HISTOGRAM_ENUM(bbb, ...)
+//    // Counters are not instantiated until incremented.
+//    EnumMap bbb;
+//    EnumMap& bbb_() { return bbb; }
+//    ...
+//  };
+struct QuicheStats {
+  QUICHE_STATS(STRUCT_MEMBER_COUNTER, STRUCT_MEMBER_GAUGE, STRUCT_MEMBER_HISTOGRAM,
+               STRUCT_MEMBER_ENUM_HISTOGRAM, STRUCT_MEMBER_STRUCT)
+};
+
+} // namespace
+
+class QuicStatsImpl {
+public:
+  void initializeStats(Envoy::Stats::Scope& scope) {
+    scope_ = &scope;
+    std::string prefix("quiche");
+    stats_.reset(
+        new QuicheStats{// // QUIC_SERVER_HISTOGRAM_COUNTER(aaa, ...)
+                        // scope.histogram("quiche.aaa", ...),
+                        //
+                        // // QUIC_SERVER_HISTOGRAM_ENUM(bbb, ...)
+                        // {},
+                        //
+                        // // QUIC_SERVER_HISTOGRAM_ENUM(a.b.c.d, ...)
+                        // // QUIC_SERVER_HISTOGRAM_ENUM(a.b.c.e, ...)
+                        // {{{ {},{} }}},
+                        QUICHE_STATS(INIT_STATS_COUNTER, INIT_STATS_GAUGE, INIT_STATS_HISTOGRAM,
+                                     INIT_STATS_ENUM_HISTOGRAM, INIT_STATS_STRUCT)});
+  }
+
+  void resetForTest() {
+    stats_.reset();
+    scope_ = nullptr;
+  }
+
+  Envoy::Stats::Counter& createCounter(std::string name) { return scope_->counterFromString(name); }
+
+  QuicheStats& stats() {
+    RELEASE_ASSERT(stats_ != nullptr, "Quiche stats is not initialized_.");
+    return *stats_;
+  }
+
+private:
+  Envoy::Stats::Scope* scope_;
+  std::unique_ptr<QuicheStats> stats_;
+};
+
+QuicStatsImpl& getQuicheStats() {
+  // Each worker thread has its own set of references to counters, gauges and
+  // histograms.
+  static thread_local QuicStatsImpl stats_impl = QuicStatsImpl();
+  return stats_impl;
+}
+
+#define QUIC_SERVER_HISTOGRAM_ENUM_IMPL(enum_name, sample, enum_size, docstring)                   \
   do {                                                                                             \
+    CounterMap& enum_map = getQuicheStats().stats().enum_name##_();                                \
+    uint32_t key = static_cast<uint32_t>(sample);                                                  \
+    auto it = enum_map.find(key);                                                                  \
+    if (it == enum_map.end()) {                                                                    \
+      auto result = enum_map.emplace(                                                              \
+          key, getQuicheStats().createCounter(absl::StrCat("quiche." #enum_name ".", sample)));    \
```

Comment:
given the above comments where `"quiche"` is captured as a StatName in your structure, you can use join() without taking. a lock. What I'm about to write is not completely coherent as I don't fully understand your structures and maps, but this is the basic pattern you want, where you join a pre-symbolized prefix to a dynamically allocated string. I'm ignoring 'sample' for the moment until I get more clarity on how this works. Ideally one of 'enum' or 'sample' should be pre-symbolized like `quiche_stat_name_`, to reduce the amount of dynamic allocation we need to do.
```
   if (it == enum_map.end()) { 
     QuicStatsImpl& quiche_stats = getQuicheStats();
     Envoy::Stats::SymbolTable& symbol_table = quiche_stats.symbolTable();
     Envoy::Stats::StatNameDynamicStorage storage(#enum_name, symbol_table);
     Envoy::Stats::SymbolTable::StoragePtr joined = symbol_table.join({
        quiche_stats.quicheStatName(), storage.statName()});
    EXPECT_EQ(absl::StrCat("a.b.", stat_str, ".c.d"), table_->toString(StatName(joined.get())));
    auto result = quiche_stats.scope().counterFromStatName(StatName(joined.get());
    ...
```

You could also wrap that in a helper method on QuicheStatsImpl, though you'd need several variants of that helper method for gauges, histograms. Maybe with some helper methods.

But the above block will not take a global lock, and it will only use memory-inefficient dynamic allocation for `#enum_name` and not for the `"quiche"` prefix.

Speaking of which, what's the threading story on your CounterMap?

Reply:
I see. Let's not re-implement source/common/common/thread_local_store.cc here. It's very expensive and not necessary to make this thread-local. A central tenant of the stats architecture is that the stat name strings when fully elaborated are incredibly expensive in memory , and we don't want to have thread-local maps of strings.

However I do think the pattern I suggest itself should be abstracted into the stats system, and that would also simplify other existing call-sites. Your PR convinces me to work on doing that now, but please don't let that block this PR. If mine comes in after yours I'll refactor this to clean it up. Here's a sample of a unit-test which I'm working on, which 
```
  ScopePtr scope = store_->createScope("scope.");
  Counter& c1 = Utility::counterFromElements(*scope, {"a", "b"});
  EXPECT_EQ("scope.a.b", c1.name());
  StatName token = pool_.add("token");
  Counter& c2 = Utility::counterFromElements(*scope, {"a", token, "b"});
  EXPECT_EQ("scope.a.token.b", c2.name());
  StatName suffix = pool_.add("suffix");
  Counter& c3 = Utility::counterFromElements(*scope, {token, suffix});
  EXPECT_EQ("scope.token.suffix", c3.name());
```
This uses the nifty `absl::variant` so we can mix pre-symbolized tokens and tokens specified via string_view.

Reply:
I have a draft PR: https://github.com/envoyproxy/envoy/pull/10735

ptal at the pattern as encapsulated there.

Reply:
> I see. Let's not re-implement source/common/common/thread_local_store.cc here. It's very expensive and not necessary to make this thread-local.  A central tenant of the stats architecture is that the stat name strings when fully elaborated are incredibly expensive in memory , and we don't want to have thread-local maps of strings.

I actually don't know ThreadLocalStore well. I think it's a totally different thing from the stats struct I created here which is merely a sets of references. The Scope instance I used in `initializeStats()` is actually supposed to be a ThreadLocalStore instance. Please let me know if my understanding is wrong.

thread-localizing QuicStatsImpl is for QUICHE code to access the stats struct instance as QUICHE doesn't have interface to pass it in. We are adding a context to QUICHE actually, but it's not for platform abstraction purpose. I don't know how the api would turn out to be and if it can be adopted by platform stuff. But for now, making QuicStatsImpl thread-local is the only way to make it accessible in QUICHE.  As to the stat name strings, I don't think they are kept anywhere in the stats struct I proposed. The struct instantiated is similar to what stats_macro.h does, with all the stats name as variable name, and the enum map is keyed on integer.

> However I do think the pattern I suggest itself should be abstracted into the stats system, and that would also simplify other existing call-sites. Your PR convinces me to work on doing that now, but please don't let that block this PR. If mine comes in after yours I'll refactor this to clean it up. Here's a sample of a unit-test which I'm working on, which
> 
> ```
>   ScopePtr scope = store_->createScope("scope.");
>   Counter& c1 = Utility::counterFromElements(*scope, {"a", "b"});
>   EXPECT_EQ("scope.a.b", c1.name());
>   StatName token = pool_.add("token");
>   Counter& c2 = Utility::counterFromElements(*scope, {"a", token, "b"});
>   EXPECT_EQ("scope.a.token.b", c2.name());
>   StatName suffix = pool_.add("suffix");
>   Counter& c3 = Utility::counterFromElements(*scope, {token, suffix});
>   EXPECT_EQ("scope.token.suffix", c3.name());
> ```
> 
> This uses the nifty `absl::variant` so we can mix pre-symbolized tokens and tokens specified via string_view.

Could you confirm that  `Envoy::Stats::StatNameDynamicStorage storage(#enum_name, symbol_table)` also works if `enum_name` is `a.b`? If not, we need to split it before creating StatNameDynamicStorage for them.

Reply:
I am against adding string-maps into thread-local storage. It's very wasteful to have copies of all these strings. I don't think you need to make these stats structs thread-local. You can use thread-safe APIs that don't take locks, based on my suggestions above. LMK if you need more help.

RE "works if enum_name is a.b": yes it does. However if either "a" or "b" can be known at compile-time it would be preferable to record a symbolized-stat on construction of your context, but please don't replicate that per thread.

Reply:
Sorry, I wasn't reading carefully enough: your CounterMap is from the enum int value not the string, so you are not duplicating the strings. My bad.

I still think the solution I propose is simpler and cleaner than having another layer of TLS-maps. Note that we already have TLS-maps in source/common/stats/thread_local_store.cc, so having TLS-maps in front of TLS-maps just seems needlessly complex.

Reply:
> I still think the solution I propose is simpler and cleaner than having another layer of TLS-maps. Note that we already have TLS-maps in source/common/stats/thread_local_store.cc, so having TLS-maps in front of TLS-maps just seems needlessly complex.

The stats struct I was trying to create is just for fast accessing those counters and histograms. Are you saying search in ThreadLocalStore via something like counterFromStatName() is efficient enough so that we don't need to keep a reference of these stats instances?  My understanding is that for the stats which are instantiated at start up via StatNameManagedStorage (so called pre-symbolized?) is not thread local so accessing them needs lock and so we want to instantiate them at start up. Is it correct?

Reply:
It's probably about the same as your thread-locked structure. You are not gaining much by putting another tls-scoped cache in front of the tls-scoped cache.